### DISCUSSION TITLE
How do I alter PanelLoaded() ?  I have a midi out "echo" bug..

### DISCUSSION THREAD

**Discussion by Synthetech** [RomanKubiak/ctrlr#212] [Q&A]
Please help me figure out how to do what they are discussing here:

https://ctrlr.org/forums/topic/midimessagereceived-executed-2-times/#post-69489

Something about altering the PanelLoaded() function.

My problem is I have doubling of midi messages out from Ctrlr when it is first opened (standalone).
I can get it to stop by deselecting an Input midi device (none) and then reselect it(my midi interface).. but it's a pain to do every time I start up Ctrlr.
It'd be nice if the panel actually could do this itself each time it opened.

The problems are noisy Note Off's and latching notes from a synth's Arpeggiator.
As said, problems go away when I cycle the Input selection just one time after first opening the standalone panel.

Any help would be greatly appreciated.
I have had minimal experience tinkering with the LUA scripting.
Zero doing full builds..  I was hoping what I am trying to do wont require a build.

Thanks!

**Comment by Ted-juh** [comment]
What Goodweather is telling you is to alter the function that is "fired" when the panel is loaded. So if you open a panel, go to editor mode (ctrl-e), click on an empty space in your panel, then in the properties on the right scroll down until you reach:

"LuaPanelLoaded" or its description "Called when the panel has finished loading".

Is there a script attached? If yes, then click on the "pen" symbol. The lua-editor will open the function. Then add the following to the function:

```
-- Reset Midi INPUT devices to avoid having MidiMessageReceived method called twice
-- If resetting Controller devices then the bidirectional behaviour is lost (most probably a bug)

	sMidiDevices = panel:getProperty("panelMidiInputDevice")
	panel:setPropertyString("panelMidiInputDevice", "-- None")
	panel:setPropertyString("panelMidiInputDevice", sMidiDevices)
```

Then in the LuaEditor, click "File" and then "Save and Compile". Save the panel as well. Close the panel, reopen the panel. And now the double midi message should be gone.

If there is no function attached, click on the symbol next to the "pen" symbol, add "init" (or "panelLoaded" or "guacemolesAreHealthy", whatever you want to name it) as a name. Then click the "pen" symbol and do the same as described above.



G'luck.

**Reply by Synthetech** [reply]
Thank you for your help DnaldeT.

I followed your instructions and while the panel did accept and compile the method, it unfortunately has not fixed the problem.

Guess I have to deal with the bug manually as I have been.

**Comment by dobo365** [comment]
goodweather here ;-)
Strange that it didn't work. It works fine for me on all my panels.
There is another double message issue but it is in VST (see posts from Possemo).

If you want, you can provide us your panel or a limited version of it where the issue is and we can have a look at it.
Another way to solve and identify issues is to restart from scratch a new panel, just add a minimum of things and see how it reacts.
In that way you can identify Ctrlr issues or your own ones ;-)

**Comment by dnaldoog** [comment]
Maybe try this?


1. In "LuaPanelLoaded" --- "Called when the panel has finished loading"  create  **init()** as below:
2. Create a **timerCallback** function.
3. Run that timer in **init()** 

<pre>
<code>
init = function(--[[ CtrlrInstance --]] type)
    panel:setPropertyInt("panelMidiPauseOut", 1)
    timer:setCallback(1, timerCallback)
    timer:startTimer(1, 500)
end

function timerCallback(timerId)
    if timerId == 1 then -- ALLOW midi to run
        panel:setPropertyInt("panelMidiPauseOut", 0)
        timer:stopTimer(timerId)
    end
end

</code>
</pre>

**Reply by Synthetech** [reply]
Thanks so much for the support here!

dnaldoog
I'll give that a try.. not sure what the timer function is for.
Is it to delay midi output just one time at the initial startup?


goodweather
I may link a download to my panel later on.
As you say, I may have somehow created the problem myself in my panel.
I probably should try to grab the latest version of Ctrlr and build a panel with only the functions I'm having a problem with and see if the issue happens again.

**Reply by dnaldoog** [reply]
> Thanks so much for the support here!
> 
> dnaldoog
> I'll give that a try.. not sure what the timer function is for.
> Is it to delay midi output just one time at the initial startup?
> 

Yes that’s correct.

Did you manage to try it and did you manage to solve the problem?

**Reply by Synthetech** [reply]
dnaldoog
Sorry I took so long to get back..

Well I did figure out what the problem was.
I had my Axiom 25 Keyboard as:
MIDI IN
MIDI Controller
MIDI OUT

I've done this all the time and did not realize **_I WAS MAKING THE PROBLEM.._**

When I only selected like this for the Axiom:
MIDI Controller
MIDI OUT

and NOT SELECT it for MIDI IN, the problem immediately went away!

So glad that's all it was.  I must have created some kind of loop even though my Axiom's MIDI IN was not connected to anything, only it's USB cord and the MIDI OUT to my custom Synth.
